storage/replica_command.go

```diff
@@ -247,6 +252,48 @@ func (r *Replica) ReverseScan(batch engine.Engine, h roachpb.BatchRequest_Header
 	return reply, intents, err
 }
 
+// BeginTransaction writes the initial transaction record. Fails in
+// the event that a transaction record is already written. This may
+// occur if a transaction is started with a batch containing writes
+// to different ranges, and the range containing the txn record fails
+// to receive the write batch before a heartbeat or txn push is
+// performed first and aborts the transaction.
+func (r *Replica) BeginTransaction(batch engine.Engine, ms *engine.MVCCStats, h roachpb.BatchRequest_Header, args roachpb.BeginTransactionRequest) (roachpb.BeginTransactionResponse, error) {
+	var reply roachpb.BeginTransactionResponse
+
+	if h.Txn == nil {
```

Comment:
we have (or should have) this and the checks below in all of the `.*Transaction` methods, I think. Consider a helper.